pretraining/utils/dataset_utils.py

```diff
@@ -1350,18 +1359,26 @@ def __init__(
                     )
                 # Grab new reader if new
                 if shard != last_shard:
+                    # Close previous file
+                    if shard_file is not None:
+                        shard_file.close()
                     path = os.path.join(datapath, dataset, shard)
-                    reader = pa.ipc.open_file(path)
+                    shard_file = open(path) # Closed below
+                    reader = pa.ipc.open_file(shard_file)
                     last_shard = shard
                     shardcount += 1
-                ndocs = reader.num_record_batches
-                doc_start = (ndocs * frag) // worldsize
-                doc_end = (ndocs * frag + ndocs) // worldsize
-                # Read into temp docset_slim
-                for i in range(doc_start, doc_end):
-                    doc = reader.get_batch(i)["tokens"]  # .to_pylist()
-                    docset_slim[(dataset, shardcount, i)] = doc
-            del reader
+                if reader is not None:
+                    ndocs = reader.num_record_batches
+                    doc_start = (ndocs * frag) // worldsize
+                    doc_end = (ndocs * frag + ndocs) // worldsize
+                    # Read into temp docset_slim
+                    for i in range(doc_start, doc_end):
+                        doc = reader.get_batch(i)["tokens"]  # .to_pylist()
+                        docset_slim[(dataset, shardcount, i)] = doc
+            if reader is not None:
+                del reader
+            if shard_file is not None:
+                shard_file.close()
```

Comment:
this looks like more than just black reformatting changes?

Reply:
I added those, because `open_file` was used with a `path` which according to the stubs does not work and also `open_file` may return None. 

I added the `shard_file` file descriptor and added the logic to close files.

Reply:
without testing, I think we need to be cautious about making any changes that aren't just mypy type hints in this PR. reading closer... this change is opening a file and creating a new file object, and introduces new logic about where it gets closed.

I just tried locally doing `f=open(some_arrow_file_path); reader = pa.ipc.open_file(f)` and got an error of `TypeError: binary file expected, got text file`.

To make this work, I think it would have needed to be `shard_file = open(path, mode='rb')`. Without actually testing it, to be on the safe side, I think it'll be better to just use a targeted type ignore hint with a comment, or fix it some simpler way.